### PR TITLE
[WIP] feat: Basic implementation enabling azure platform

### PR DESCRIPTION
**Pull request by serbrech**
This PR is being split into smaller chunks and should not be merged.

Basic implementation of the platform specific code in the installer to enable Azure as a target platform.

It is bypassing the `MachineSpecProvider` until we have it ready. 
There are a few shortcuts, but it allows to run the installer, generate all the assets and the tfvars needed for the terraform PR (#1454) to deploy a cluster on Azure.

The aim for this PR is to get a minimum working solution to be able to iterate in smaller subsequent PRs.

TODO : 

- Use  https://github.com/openshift/cluster-api-provider-azure
- Import azure terraform and required plugins
- Remove hardcoded values

### CODE REVIEW

**Comment by serbrech** [comment]
@enxebre @imcleod @juan-lee @awesomenix

File: pkg/asset/installconfig/installconfig.go [hunk 1]
```diff
@@ -4,15 +4,14 @@ import (
 	"os"
 
 	"github.com/ghodss/yaml"
-	"github.com/pkg/errors"
```
Comment (abhinavdahiya) [inline pkg/asset/installconfig/installconfig.go]: nit: keep the imports as-is.

File: pkg/asset/installconfig/platform.go [hunk 2]
```diff
@@ -4,18 +4,19 @@ import (
 	"fmt"
 	"sort"
 
-	"github.com/pkg/errors"
```
Comment (abhinavdahiya) [inline pkg/asset/installconfig/platform.go]: nit: keep the imports as-is

File: pkg/asset/manifests/dns.go [hunk 3]
```diff
@@ -6,18 +6,17 @@ import (
 	"strings"
 
 	"github.com/ghodss/yaml"
-	"github.com/pkg/errors"
```
Comment (abhinavdahiya) [inline pkg/asset/manifests/dns.go]: nit: keep the imports as-is

File: pkg/asset/installconfig/installconfig.go (hunk 1 above)
Comment (serbrech) [reply pkg/asset/installconfig/installconfig.go]: You don't use `gofmt`? my editor runs it at every file save, so I didn't think that would be a problem.
Happy to oblige, what's your convention?

Comment (serbrech) [reply pkg/asset/installconfig/installconfig.go]: aha, extra line disappeared 👍fixing it.

File: pkg/types/installconfig.go [hunk 4]
```diff
@@ -91,6 +93,10 @@ type Platform struct {
 	// OpenStack is the configuration used when installing on OpenStack.
 	// +optional
 	OpenStack *openstack.Platform `json:"openstack,omitempty"`
+
+	// OpenStack is the configuration used when installing on OpenStack.
```
Comment (serbrech) [inline pkg/types/installconfig.go]: s/OpenStack/Azure

File: pkg/types/azure/platform.go [hunk 5]
```diff
@@ -0,0 +1,18 @@
+package azure
+
+// Platform stores all the global configuration that all machinesets
+// use.
+type Platform struct {
+	// Region specifies the AWS region where the cluster will be created.
```
Comment (enxebre) [inline pkg/types/azure/platform.go]: nit: s/aws/azure

File: pkg/types/azure/validation/platform_test.go [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+		},
+		{
+			name: "invalid region",
+			platform: &azure.Platform{
+				Region: "bad-region",
+			},
+			valid: false,
+		},
+		{
+			name: "valid machine pool",
```
Comment (enxebre) [inline pkg/types/azure/validation/platform_test.go]: valid and invalid pool look the same here?

File: vendor/sigs.k8s.io/cluster-api-provider-azure/LICENSE [hunk 7]
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
```
Comment (enxebre) [inline vendor/sigs.k8s.io/cluster-api-provider-azure/LICENSE]: do we need this yet? Is go I can't find the Gopkg.* changes

File: pkg/types/azure/validation/platform.go [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+			validValues[i] = r
+			i++
+		}
+		sort.Strings(validValues)
+		return validValues
+	}()
+)
+
+// ValidatePlatform checks that the specified platform is valid.
+func ValidatePlatform(p *azure.Platform, fldPath *field.Path) field.ErrorList {
```
Comment (enxebre) [inline pkg/types/azure/validation/platform.go]: this is implemented but not called, we'll need to add it to `validatePlatform` here https://github.com/openshift/installer/blob/master/pkg/types/validation/installconfig.go#L192-L221

Comment (enxebre) [reply pkg/types/azure/validation/platform.go]: Also these error messages need to be updated:
```
    --- FAIL: TestValidateInstallConfig/missing_platform (0.00s)
    	installconfig_test.go:441: 
    			Error Trace:	installconfig_test.go:441
    			Error:      	Expect "platform: Invalid value: types.Platform{AWS:(*aws.Platform)(nil), Libvirt:(*libvirt.Platform)(nil), None:(*none.Platform)(nil), OpenStack:(*openstack.Platform)(nil), Azure:(*azure.Platform)(nil)}: must specify one of the platforms (aws, azure, none, openstack)" to match "^platform: Invalid value: types\.Platform{AWS:\(\*aws\.Platform\)\(nil\), Libvirt:\(\*libvirt\.Platform\)\(nil\), None:\(\*none\.Platform\)\(nil\), OpenStack:\(\*openstack\.Platform\)\(nil\)}: must specify one of the platforms \(aws, none, openstack\)$"
    			Test:       	TestValidateInstallConfig/missing_platform
```

File: pkg/asset/installconfig/platform.go [hunk 9]
```diff
@@ -24,7 +24,9 @@ import (
 
 // Platform is an asset that queries the user for the platform on which to install
 // the cluster.
-type platform types.Platform
+type platform struct {
```
Comment (abhinavdahiya) [inline pkg/asset/installconfig/platform.go]: Why do you need this change?

File: pkg/types/azure/dns.go [hunk 10]
```diff
@@ -0,0 +1,25 @@
+package azure
```
Comment (abhinavdahiya) [inline pkg/types/azure/dns.go]: `types` is installer's public api. therefore it has `InstallConfig` and `ClusterMetadata` please move this to `pkg/asset`

File: pkg/dns/dns.go [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+	"github.com/openshift/installer/pkg/types/aws"
+	"github.com/openshift/installer/pkg/types/azure"
+	"github.com/openshift/installer/pkg/types/libvirt"
+	"github.com/openshift/installer/pkg/types/none"
+	"github.com/openshift/installer/pkg/types/openstack"
+)
+
+//ConfigProvider is an interface that provides means to fetch the DNS settings
+type ConfigProvider interface {
+	GetBaseDomain() (string, error)
```
Comment (abhinavdahiya) [inline pkg/dns/dns.go]: this is going to be used by `pkg/asset/installconfig` only please move it there.

File: pkg/dns/dns.go [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+	GetPublicZone(name string) string
```
Comment (abhinavdahiya) [inline pkg/dns/dns.go]: Even this is going to be used by assets that use installconfig asset as input to perform base_domain -> public_zone. so it makes sense to keep them in `pkg/asset/installconfig`

File: pkg/asset/cluster/azure/azure.go [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+)
+
+// Metadata converts an install configuration to Azure metadata.
+func Metadata(clusterID, infraID string, config *types.InstallConfig) *azure.Metadata {
+	return &azure.Metadata{
+		Region: config.Platform.Azure.Region,
+		Identifier: []map[string]string{{
+			fmt.Sprintf("kubernetes.io.cluster.%s", infraID): "owned",
+		}, {
+			"openshiftClusterID": clusterID,
```
Comment (abhinavdahiya) [inline pkg/asset/cluster/azure/azure.go]: TODO: this might not be required.

File: pkg/types/azure/validation/platform_test.go (hunk 6 above)
Comment (serbrech) [reply pkg/types/azure/validation/platform_test.go]: fixed the test, and added a TODO in the machinepool validation, which is not validating anything just yet. https://github.com/openshift/installer/pull/1460/commits/fbdbf5cc1e877b3a377367e44ddb9d333e68f169

File: pkg/types/azure/dns.go (hunk 10 above)
Comment (serbrech) [reply pkg/types/azure/dns.go]: moved

File: pkg/dns/dns.go (hunk 11 above)
Comment (serbrech) [reply pkg/dns/dns.go]: moved

File: pkg/dns/dns.go (hunk 12 above)
Comment (serbrech) [reply pkg/dns/dns.go]: moved

File: pkg/asset/installconfig/platform.go (hunk 9 above)
Comment (serbrech) [reply pkg/asset/installconfig/platform.go]: allows to get the platform name without depending on more assets : https://github.com/openshift/installer/pull/1460/files?file-filters%5B%5D=.go#diff-4d231774d44727d779db320e7b6c189eR39

File: vendor/sigs.k8s.io/cluster-api-provider-azure/LICENSE (hunk 7 above)
Comment (serbrech) [reply vendor/sigs.k8s.io/cluster-api-provider-azure/LICENSE]: it was initially faked by adding a dummy implementation in the vendor.
it's dirty, but we can soon remove it.

File: pkg/asset/installconfig/platform.go (hunk 9 above)
Comment (abhinavdahiya) [reply pkg/asset/installconfig/platform.go]: > allows to get the platform name without depending on more assets

:confused: do you mind expanding on that?
in the current usage, why not call `platform.Name()` directly compared to `CurrentPlatformName` ?

Comment (serbrech) [reply pkg/asset/installconfig/platform.go]: that gives returns the name of the asset. not the name of the platform.

Comment (serbrech) [reply pkg/asset/installconfig/platform.go]: https://github.com/openshift/installer/blob/0be524e5a2d67459bc0ee6e2614b6b257c213de0/pkg/asset/installconfig/platform.go#L81

**Comment by serbrech** [comment]
I can't reconciliate the dependencies between `pkg/terraform/exec/vendor` and the root `vendor` directory.
Building works, but at runtime, I get a panic : 

```
panic: /debug/requests is already registered. You may have two independent copies of golang.org/x/net/trace in your binary, trying to maintain separate state. This may involve a vendored copy of golang.org/x/net/trace.

goroutine 1 [running]:
github.com/openshift/installer/pkg/terraform/exec/vendor/golang.org/x/net/trace.init.0()
	/Users/stephane/go/src/github.com/openshift/installer/pkg/terraform/exec/vendor/golang.org/x/net/trace/trace.go:117 +0x16c
```

[I tried to override](https://github.com/openshift/installer/blob/0be524e5a2d67459bc0ee6e2614b6b257c213de0/pkg/terraform/exec/Gopkg.toml#L30-L32) the `golang.org/x/net` package so that dep would see it's already there, as this seems to be the way you're going about this, but dep still import the net package twice. 
Can you please try to fix this? 

IMO, this nested dep situation is really risky and brittle.[It is known to cause problem too](https://github.com/golang/dep/issues/985#issuecomment-321353370). What's the reason behind it? why not just a single root vendor?

**Comment by abhinavdahiya** [comment]
> IMO, this nested dep situation is really risky and brittle.[It is known to cause problem too](https://github.com/golang/dep/issues/985#issuecomment-321353370). What's the reason behind it? why not just a single root vendor?

keeping single vendor with terraform and terraform providers was difficult because of many reasons like:
terraform has dependencies on pretty old libs which is in-compatible with newer versions like the
[azure-sdk](https://github.com/hashicorp/terraform/blob/d48069534832fdb09919a5fe46500755540cbf0a/go.mod#L5)
terraform also includes terraform providers like [aws](https://github.com/hashicorp/terraform/blob/d48069534832fdb09919a5fe46500755540cbf0a/go.mod#L125) that prevented us from moving to newer cloud sdk libraries.
and we can try to flatten out when `0.12` terraform lands, as it uses newer libraries..

But it might still be desire-able to keep terraform and terraform-providers separate because it allows installer to use newer versions of libs that might not allowed due to vendored terraform and terraform-providers...

File: pkg/asset/cluster/azure/azure.go (hunk 13 above)
Comment (serbrech) [reply pkg/asset/cluster/azure/azure.go]: azure doesn't support `/` in tags, and I don't know what these tags are used for in openshift.
to get the terraform deployment to work, I changed the `/` with a `.`.
if the `kubernetes/io/cluster' tags  are not needed for openshift, I'm happy to remove them

File: pkg/types/azure/validation/platform.go (hunk 8 above)
Comment (serbrech) [reply pkg/types/azure/validation/platform.go]: both were addressed

**Comment by serbrech** [comment]
> But it might still be desire-able to keep terraform and terraform-providers separate because it allows installer to use newer versions of libs that might not allowed due to vendored terraform and terraform-providers...

@abhinavdahiya Are you able to remediate the dependency conflict I am facing then?

**Comment by abhinavdahiya** [comment]
> > But it might still be desire-able to keep terraform and terraform-providers separate because it allows installer to use newer versions of libs that might not allowed due to vendored terraform and terraform-providers...
> 
> @abhinavdahiya Are you able to remediate the dependency conflict I am facing then?

did this patch not help adding the azure terraform plugin?

```patch
From 7b71c902cdcebfda0e06fc57f379732a6505069e Mon Sep 17 00:00:00 2001
From: Abhinav Dahiya <abhinav.dahiya@redhat.com>
Date: Tue, 26 Mar 2019 16:32:29 -0700
Subject: [PATCH] terraform/plugins: add terraform-provider-azurerm to internal
 plugins

- Add the `azure.go` that installs terraform-provider-azurerm to map of KnownPlugins.

- Add "golang.org/x/net/trace" to list of overrides in `Gopkg.toml` because only single `net/trace` handler can be rgistered and vendored `terraform` already registers handlers.


panic: /debug/requests is already registered. You may have two independent copies of golang.org/x/net/trace in your binary, trying to maintain separate state. This may involve a vendored copy of golang.org/x/net/trace.

goroutine 1 [running]:
github.com/openshift/installer/pkg/terraform/exec/plugins/vendor/golang.org/x/net/trace.init.0()
        /home/adahiya/go/src/github.com/openshift/installer/pkg/terraform/exec/plugins/vendor/golang.org/x/net/trace/trace.go:123 +0x17e

- Add constraint for `github.com/terraform-providers/terraform-provider-azurerm` to `Gopkg.toml`

- Run `dep ensure`
---
 pkg/terraform/exec/plugins/Gopkg.toml |  7 ++++++-
 pkg/terraform/exec/plugins/azure.go   | 15 +++++++++++++++
 2 files changed, 21 insertions(+), 1 deletion(-)
 create mode 100644 pkg/terraform/exec/plugins/azure.go

diff --git a/pkg/terraform/exec/plugins/Gopkg.toml b/pkg/terraform/exec/plugins/Gopkg.toml
index cabffed83..3e8f80ce0 100644
--- a/pkg/terraform/exec/plugins/Gopkg.toml
+++ b/pkg/terraform/exec/plugins/Gopkg.toml
@@ -1,6 +1,7 @@
 ignored = [
   "github.com/libvirt/libvirt-go",
-  "github.com/hashicorp/terraform/*"
+  "github.com/hashicorp/terraform/*",
+  "golang.org/x/net/trace"
 ]
 
 [prune]
@@ -16,6 +17,10 @@ ignored = [
   name = "github.com/terraform-providers/terraform-provider-aws"
   version = "1.52.0"
 
+[[constraint]]
+  name = "github.com/terraform-providers/terraform-provider-azurerm"
+  version = "1.23.0"
+
 [[constraint]]
   name = "github.com/terraform-providers/terraform-provider-ignition"
   version = "1.0.1"
diff --git a/pkg/terraform/exec/plugins/azure.go b/pkg/terraform/exec/plugins/azure.go
new file mode 100644
index 000000000..fe3968415
--- /dev/null
+++ b/pkg/terraform/exec/plugins/azure.go
@@ -0,0 +1,15 @@
+package plugins
+
+import (
+	"github.com/hashicorp/terraform/plugin"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm"
+)
+
+func init() {
+	exec := func() {
+		plugin.Serve(&plugin.ServeOpts{
+			ProviderFunc: azurerm.Provider,
+		})
+	}
+	KnownPlugins["terraform-provider-azurerm"] = exec
+}
```

Or are you seeing error when adding something else to the vendor?

**Comment by serbrech** [comment]
@abhinavdahiya I'm not adding the terraform plugin, I am adding autorest which is a prerequisite to azure-sdk-go for the installer to query azure during configuration.

**Comment by serbrech** [comment]
After trial and error, we figured out that for the `[[override]]` statement to be taken into account by `dep` under `pkg/terraform/exec`, we had to manually remove the dependency from the lock file before running `dep ensure`. 

However, the terraform package uses `golang.io/x/net/html/charset`, which is not used by anything else, and thus, not imported into the root `Gopkg.lock`.

To resolve this, we had to add a `require = ["golang.io/x/net/html/charset"]` to force it as a root dependency.

**Comment by serbrech** [comment]
rebased on #1534

**Comment by abhinavdahiya** [comment]
this was replaced with multiple PRs
like https://github.com/openshift/installer/pull/1721 https://github.com/openshift/installer/pull/1454 etc.

/close

**Pull request closed without merging**